### DISCUSSION TITLE
v2.31.0

### DISCUSSION THREAD

**Discussion by joseluisq** [static-web-server/static-web-server#445] [Announcements]
This new `v2.31.0` release brings several bug fixes, features like a new compression level option, logs for virtual hosts, better-accepted encodings handling for pre-compressed files and other improvements.

**Fixes**

- d22e2dd Bugfix/security dependency updates including bytes, serde, toml, async-compression, flate2, brotli and other crates (also ea96328).
- 1aae13c Directory Listing: Empty file sizes are displayed incorrectly. PR #385 by @palant.
- 25c171b Directory Listing: File sizes are wrongly displayed in decimal format. PR #376 by @miroim.
- e7bfaa2 Lib: Compile errors if only `deflate` compression is enabled. PR #383 by @palant.
- 195f706 Docker: Unsupported system page size when using Linux ARM64 musl (E.g. Raspberry Pi 5). PR #443.
- d4046d9 Unexpected `cfg` condition-name `wasm` in Rust nightly. PR #441.
- 1b13a74 Misc: Clippy warnings caused by test code. PR #382 by @palant.
- 0792606 CI: Code analysis workflow.
- 936b224 CI: Prevent running the FreeBSD `devel` pipeline unnecessarily.

**Refactorings**

- 85e3da Lib: Make request `body` type generic across modules. PR #375 by @palant.
- 18f550a Directory Listing: Drop custom type conversion between `SystemTime` and `DateTime`. PR #384 by @palant.
- 219ec6c Move Remote and Real IP addresses logging to a new module. PR #388.
- cfd8bb1 Re-organize file system-related modules. PR #424.
- 7cf72e6 Misc: Several project config and doc file improvements.

**Features**

- 9cbf95b Compression level option support. PR #381 by @palant. See docs.
- 778477e Log virtual-hosts information per request. PR #442.
- 7f59da9 Look for all accepted encodings for pre-compressed files but keep the quality order. PR #439 by @kobutri.
- d2eaa74 Benchmark: Add Caddy server to performance benchmarks. PR #379 by @palant.
- 2faeef5 Show detailed information about the server via `--version` (`-V`) flag. PR #444.

For more details see the [v2.31.0 milestone](https://github.com/static-web-server/static-web-server/milestone/18?closed=1) and the full changelog [v2.30.0...v2.31.0](https://github.com/static-web-server/static-web-server/compare/v2.30.0...v2.31.0).

<hr /><em>This discussion was created from the release <a href='https://github.com/static-web-server/static-web-server/releases/tag/v2.31.0'>v2.31.0</a>.</em>

**Comment by sergeevabc** [comment]
2.31.0 stopped working on Windows 7, because Rust language developers betrayed Windows 7 users by removing the feature to build binaries for this OS. The current workaround is to set a legacy compilation chain like @ozwaldorf [did with Lutgen-RS][1]. Please, do.

[1]: https://github.com/ozwaldorf/lutgen-rs/commit/ccdc1e3892f2f96b5ac595ec3aed662788e65960

**Comment by joseluisq** [comment]
> 2.31.0 stopped working on Windows 7, because Rust language developers betrayed Windows 7 users by removing the feature to build binaries for this OS. The current workaround is to set a legacy compilation chain like @ozwaldorf [did with Lutgen-RS](https://github.com/ozwaldorf/lutgen-rs/commit/ccdc1e3892f2f96b5ac595ec3aed662788e65960). Please, do.

Unfortunately, it just skipped from my radar but I'm fine with sticking with `1.77.2` for a while just for `x86_64-pc-windows-msvc`, `i686-pc-windows-msvc` and `x86_64-pc-windows-gnu` only.

However, keep in mind that [MS already stopped supporting Windows 7](https://learn.microsoft.com/en-us/lifecycle/products/windows-7)  (also for extended users until the beginning of 2023).
So I don't know how long we should keep supporting an unmaintained platform. Maybe until next year give users time to update their platforms and also provide them instructions to build SWS from the source for that platform (which should be trivial). Our MSRV is 1.74.0 so it should not be a problem for them.

Please file an issue and let me know your thoughts on this.

**Reply by joseluisq** [reply]
@sergeevabc FYI https://github.com/static-web-server/static-web-server/pull/447
